src/cmap/wait_queue.rs

```diff
@@ -1,177 +1,85 @@
+        let future = self.semaphore.acquire(1);
 
-        let condvar = Arc::new(Condvar::new());
-        guard.queue.push_back(condvar.clone());
-
-        // If we're already at the front of the wait queue, we don't need to wait. We don't pop
-        // ourself off of the wait queue yet though, since another thread may enter while we're
-        // waiting for a connection to become available.
-        if guard.queue.len() == 1 {
-            return Ok(WaitQueueHandle::new(guard, condvar));
-        }
-
-        let guard = if let Some(timeout) = guard.timeout {
-            // Wait until all of the WaitQueueHandles in the queue in front of the current thread's
-            // are dropped or a timeout occurs.
-            let (guard, result) = condvar.wait_timeout(guard, timeout).unwrap();
-
-            if result.timed_out() {
-                return Err(ErrorKind::WaitQueueTimeoutError {
-                    address: guard.address.clone(),
+        let releaser = if let Some(timeout) = self.timeout {
+            RUNTIME.await_with_timeout(Box::pin(future), timeout).await.map_err(|_| {
+                ErrorKind::WaitQueueTimeoutError {
+                    address: self.address.clone(),
                 }
-                .into());
-            }
-
-            guard
+            })?
         } else {
-            // Wait until all of the WaitQueueHandles in the queue in front of the current thread's
-            // are dropped.
-            condvar.wait(guard).unwrap()
+            future.await
         };
 
-        Ok(WaitQueueHandle::new(guard, condvar))
+        Ok(WaitQueueHandle {
+            semaphore_releaser: releaser
+        })
     }
 
-    /// Notifies the wait queue that the next thread in the queue can be woken up.
-    pub(super) fn notify_ready(&self) {
-        self.inner.lock().unwrap().notify_ready();
+    /// Signal that the front of the queue (if there is one) is ready to wake up.
+    pub(super) fn wake_front(&self) {
+        if self.semaphore.permits() >= self.max_handles {
+            panic!("greater than {} connections checked back into pool with address {}",
+                   self.max_handles,
+                   self.address.clone()
+            );
+        }
+        self.semaphore.release(1);
     }
 }
 
-impl WaitQueueInner {
-    /// Notifies the wait queue that the next thread in the queue can be woken up. This is
-    /// explicitly defined on `WaitQueueInner` so that `WaitQueueHandle::drop` can call it.
-    pub(super) fn notify_ready(&self) {
-        if let Some(condvar) = self.queue.front() {
-            condvar.notify_one();
-        }
+/// A handle to a `WaitQueue` that will wake up the front of the queue when dropped.
+/// To disable this behavior, call `WaitQueueHandle::disarm`.
+pub(super) struct WaitQueueHandle<'a> {
+    semaphore_releaser: SemaphoreReleaser<'a>
+}
+
+impl<'a> WaitQueueHandle<'a> {
+    pub(super) fn disarm(&mut self) {
```

Comment:
Is there a reason we can't just call `disarm` from a `Drop` implementation on `WaitQueueHandle`? It's an `async` function, so I don't think there should be any issue.

Reply:
We don't want the handle to be disarmed in drop as that would defeat its purpose. This handle is useful for scenarios where you get to the front of the queue but fail to create / acquire a connection and return early via `?`. That way, instead of waking the front on every early return, we only have to disarm before we return successfully, and the queue will otherwise be woken when the handle drops.